doc/man/nfsref.8
.\"@(#)nfsref.8"
.\"
.\" @file doc/man/nfsref.8
.\" @brief man page for nfsref command
.\"

.\"
.\" Copyright 2011 Oracle.  All rights reserved.
.\"
.\" This file is part of fedfs-utils.
.\"
.\" fedfs-utils is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License version 2.0 as
.\" published by the Free Software Foundation.
.\"
.\" fedfs-utils is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License version 2.0 for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" version 2.0 along with fedfs-utils.  If not, see:
.\"
.\"	http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt
.\"
.TH NFSREF 8 "@publication-date@"
.SH NAME
nfsref \- manage NFS referrals
.SH SYNOPSIS
.B nfsref
.RB [ \-?d ]
.RB [ \-t
.IB type ]
.B add
.I pathname server export
.RI [ " server"
.IR export " ... ]"
.P
.B nfsref
.RB [ \-?d ]
.RB [ \-t
.IB type ]
.B remove
.I pathname
.P
.B nfsref
.RB [ \-?d ]
.RB [ \-t
.IB type ]
.B lookup
.I pathname
.SH INTRODUCTION
NFS version 4 introduces the concept of
.I file system referrals
to NFS.
A file system referral is like a symbolic link on a file server
to another file system share, possibly on another file server.
On an NFS client, a referral behaves like an automounted directory.
The client, under the server's direction, mounts a new NFS export
automatically when an application first accesses that directory.
.P
Referrals are typically used to construct a single file name space
across multiple file servers.
Because file servers control the shape of the name space,
no client configuration is required,
and all clients see the same referral information.
.P
The Linux NFS server supports NFS version 4 referrals.
Administrators can specify the
.B refer=
export option in
.I /etc/exports
to configure a list of exports from which the client can choose.
See
.BR exports (5)
for details.
.P
The
.BR nfsref (8)
command provides an alternate way to configure NFS referrals.
This command stores referral information
as metadata within a leaf directory in an exported file system.
The metadata it stores can contain one of two types of information:
.IP "\fIA list of Fileset Locations\fP"
A set of server name and export path pairs which are returned
verbatim to clients during an NFS referral event.
This is known as an
.IR "NFS basic junction" .
.IP "\fIA Fileset Name\fP"
The name of an LDAP record which contains information to return
to clients during an NFS referral event.
This is known as a
.IR "FedFS junction" .
.P
A directory can hold either an NFS basic junction or a FedFS junction,
but not both.
When a directory acts as a junction, its regular contents remain,
but are no longer visible to NFS clients.
.P
By storing the location information in an LDAP directory,
FedFS junctions on multiple file servers can refer to
the same copy of location information.
This common locations metadata can be updated
via a single administrative operation,
altering the file name space consistently across all servers.
The
.BR fedfs (7)
man page has more information.
.SH DESCRIPTION
The
.BR nfsref (8)
command is a simple way to get started managing junction metadata.
Other administrative commands provide richer access to junction information.
.SS Subcommands
Valid
.BR nfsref (8)
subcommands are:
.IP "\fBadd\fP"
Adds junction information to the directory named by
.IR pathname .
The named directory must already exist,
and must not already contain junction information.
Regular directory contents are obscured to NFS clients by this operation.
.IP
A list of one or more file server and export path pairs
is also specified on the command line.
When creating an NFS basic junction, this list is
stored in an extended attribute of the directory.
.IP
When creating a FedFS junction, FedFS records containing the
file server and export path pairs are created on an LDAP server,
and a pointer to the new FedFS records is
stored in an extended attribute of the directory.
Fresh FSN and FSL UUIDs are generated during this operation.
.IP
If junction creation is successful, the
.BR nfsref (8)
command flushes the kernel's export cache
to remove previously cached junction information.
.IP "\fBremove\fP"
Removes junction information from the directory named by
.IR pathname .
The named directory must exist,
and must contain junction information.
Regular directory contents are made visible to NFS clients again by this operation.
.IP
When removing a FedFS junction, the
.BR nfsref (8)
command also removes FSN and FSL records referred to in the junction.
.IP
If junction deletion is successful, the
.BR nfsref (8)
command flushes the kernel's export cache
to remove previously cached junction information.
.IP "\fBlookup\fP"
Displays junction information stored in the directory named by
.IR pathname .
The named directory must exist,
and must contain junction information.
.IP
When looking up an NFS basic junction, the junction information
in the directory is listed on
.IR stdout .
When looking up a FedFS junction, junction information is
retrieved from the LDAP server listed in the junction
and listed on
.IR stdout .
.P
When creating a new FedFS junction, the
.BR nfsref (8)
command reads the following environment variables:
.IP "\fBFEDFS_NSDB_HOST\fP"
Specifies the hostname of the LDAP server where new FedFS records
should reside.  If this variable is not set, the
.BR nfsref (8)
command fails.
The LDAP server specified by this variable
must be registered with the local NSDB connection
parameter database before the
.BR nfsref (8)
command can communicate with it.  See
.BR nsdbparams (8)
for more information.
.IP "\fBFEDFS_NSDB_PORT\fP"
Specifies the IP port of the LDAP server where new FedFS records
should reside.  The default value if this variable is not set is 389.
.IP "\fBFEDFS_NSDB_NCE\fP"
Specifies the distinguished name of the NSDB Container Entry
under which new FedFS records should reside.
If this variable is not set, the local NSDB connection parameter
database is searched for a default NCE for the hostname specified by
.BR FEDFS_NSDB_HOST .
If neither of these is specified, the
.BR nfsref (8)
command fails.
.IP "\fBFEDFS_NSDB_ADMIN\fP"
Specifies a distinguished name of an entity used to bind
to the LDAP server where new FedFS records should reside.
If this variable is not set, the local NSDB connection parameter
database is searched for a default bind DN for the hostname
specified by
.BR FEDFS_NSDB_HOST .
If neither of these is specified, or if this entity does not have
permission to modify the LDAP server's DIT, the
.BR nfsref (8)
command fails.
.SS Command line options
.IP "\fB\-d, \-\-debug"
Enables debugging messages during operation.
.IP "\fB\-t, \-\-type=\fIjunction-type\fP"
Specifies the junction type for the operation.  Valid values for
.I junction-type
are
.B nfs-basic
or
.BR nfs-fedfs .
.IP
For the
.B add
subcommand, the default value if this option is not specified is
.BR nfs-basic .
For the
.B remove
and
.B lookup
subcommands, the
.B \-\-type
option is not required.  The
.BR nfsref (8)
command operates on whatever junction contents are available.
.SH EXAMPLES
Suppose you have two file servers,
.I top.example.net
and
.IR home.example.net .
You want all your clients to mount
.I top.example.net:/
and then see the files under
.I home.example.net:/
automatically in
.IR top:/home .
.P
On
.IR top.example.net ,
you might issue this command as root:
.RS
.sp
# mkdir /home
.br
# nfsref --type=nfs-basic add /home home.example.net /
.br
Created junction /home.
.sp
.RE
.SH FILES
.TP
.I /etc/exports
NFS server export table
.SH "SEE ALSO"
.BR fedfs (7),
.BR nsdbparams (8),
.BR exports (5)
.sp
RFC 3530 for a description of NFS version 4 referrals
.sp
RFC 5716 for FedFS requirements and overview
.SH COLOPHON
This page is part of the fedfs-utils package.
A description of the project and information about reporting bugs
can be found at
.IR http://wiki.linux-nfs.org/wiki/index.php/FedFsUtilsProject .
.SH "AUTHOR"
Chuck Lever <chuck.lever@oracle.com>


src/nfsref/nfsref.c
/**
 * @file src/nfsref/nfsref.c
 * @brief Manage NFS referrals
 */

/*
 * Copyright 2011 Oracle.  All rights reserved.
 *
 * This file is part of fedfs-utils.
 *
 * fedfs-utils is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2.0 as
 * published by the Free Software Foundation.
 *
 * fedfs-utils is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License version 2.0 for more details.
 *
 * You should have received a copy of the GNU General Public License
 * version 2.0 along with fedfs-utils.  If not, see:
 *
 *	http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt
 */

#include <sys/types.h>
#include <sys/capability.h>
#include <sys/prctl.h>
#include <sys/stat.h>

#include <stdbool.h>
#include <stdint.h>
#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>
#include <fcntl.h>
#include <getopt.h>
#include <time.h>

#include <locale.h>
#include <langinfo.h>

#include "fedfs.h"
#include "nsdb.h"
#include "junction.h"
#include "xlog.h"
#include "gpl-boiler.h"
#include "nfsref.h"

/**
 * Short form command line options
 */
static const char nfsref_opts[] = "?dt:";

/**
 * Long form command line options
 */
static const struct option nfsref_longopts[] = {
	{ "debug", 0, NULL, 'd', },
	{ "help", 0, NULL, '?', },
	{ "type", 1, NULL, 't', },
	{ NULL, 0, NULL, 0, },
};

/**
 * Display program synopsis
 *
 * @param progname NUL-terminated C string containing name of program
 */
static void
nfsref_usage(const char *progname)
{
	fprintf(stderr, "\n%s: version " VERSION "\n", progname);
	fprintf(stderr, "Usage: %s [ -t type ] SUBCOMMAND [ ARGUMENTS ]\n\n",
		progname);

	fprintf(stderr, "SUBCOMMAND is one of:\n");
	fprintf(stderr, "\tadd        Add a new junction\n");
	fprintf(stderr, "\tremove     Remove an existing junction\n");
	fprintf(stderr, "\tlookup     Enumerate a junction\n");

	fprintf(stderr, "\nUse \"%s SUBCOMMAND -?\" for details.\n", progname);

	fprintf(stderr, "%s", fedfs_gpl_boilerplate);
}

/**
 * Program entry point
 *
 * @param argc count of command line arguments
 * @param argv array of NUL-terminated C strings containing command line arguments
 * @return program exit status
 */
int
main(int argc, char **argv)
{
	char *progname, *subcommand, *junct_path;
	enum nfsref_type type;
	int arg, exit_status;

	(void)umask(S_IWGRP | S_IWOTH);

	exit_status = EXIT_FAILURE;

	/* Ensure UTF-8 strings can be handled transparently */
	if (setlocale(LC_CTYPE, "") == NULL ||
	    strcmp(nl_langinfo(CODESET), "UTF-8") != 0) {
		fprintf(stderr, "Failed to set locale and langinfo\n");
		goto out;
	}

	/* Set the basename */
	if ((progname = strrchr(argv[0], '/')) != NULL)
		progname++;
	else
		progname = argv[0];

	xlog_stderr(1);
	xlog_syslog(0);
	xlog_open(progname);

	if (argc < 2) {
		nfsref_usage(progname);
		goto out;
	}

	type = NFSREF_TYPE_UNSPECIFIED;
	while ((arg = getopt_long(argc, argv, nfsref_opts,
			nfsref_longopts, NULL)) != -1) {
		switch (arg) {
		case 'd':
			xlog_config(D_ALL, 1);
			break;
		case 't':
			if (strcmp(optarg, "nfs-basic") == 0)
				type = NFSREF_TYPE_NFS_BASIC;
			else if (strcmp(optarg, "nfs-fedfs") == 0)
				type = NFSREF_TYPE_NFS_FEDFS;
			else {
				fprintf(stderr,
					"Unrecognized junction type: %s",
					optarg);
				exit(EXIT_FAILURE);
			}
			break;
		case '?':
			nfsref_usage(progname);
			exit(EXIT_SUCCESS);
		}
	}
	if (argc < optind + 2) {
		fprintf(stderr, "Not enough positional parameters\n");
		nfsref_usage(progname);
		goto out;
	}

	if (geteuid() != 0) {
		fprintf(stderr, "Root permission is required\n");
		goto out;
	}

	subcommand = argv[optind];
	junct_path = argv[optind + 1];

	if (strcasecmp(subcommand, "add") == 0) {
		if (argc < optind + 3) {
			fprintf(stderr, "Not enough positional parameters\n");
			nfsref_usage(progname);
			goto out;
		}
		exit_status = nfsref_add(type, junct_path, argv, optind);
		if (exit_status == EXIT_SUCCESS)
			(void)junction_flush_exports_cache();
	} else if (strcasecmp(subcommand, "remove") == 0) {
		exit_status = nfsref_remove(type, junct_path);
		if (exit_status == EXIT_SUCCESS)
			(void)junction_flush_exports_cache();
	} else if (strcasecmp(subcommand, "lookup") == 0)
		exit_status = nfsref_lookup(type, junct_path);
	else {
		xlog(L_ERROR, "Unrecognized subcommand: %s", subcommand);
		nfsref_usage(progname);
	}

out:
	exit(exit_status);
}
